test-data/unit/check-classes.test

```diff
@@ -5118,6 +5131,18 @@ class C:
 [builtins fixtures/property.pyi]
 [out]
 
+[case testClassMethodBeforeInit]
+class Foo(object):
+    @classmethod
+    def bar(cls):
+        # type: () -> Foo
+        return cls("bar")
+
+    def __init__(self, baz):
+        # type: (str) -> None
+        self.baz = baz
+[builtins fixtures/classmethod.pyi]
```

Comment:
I would add two more test cases like this:
* One where `__init__` has a forward reference in its signature
* Another with an overloaded classmethod